src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -258,51 +258,91 @@ import scala.runtime.Statics
   *  @define evaluatesAllElements This method evaluates all elements of the collection.
   */
 @SerialVersionUID(3L)
-final class LazyList[+A] private(private[this] var lazyState: () => LazyList.State[A])
+final class LazyList[+A] private (lazyState: () => LazyList[A])
   extends AbstractSeq[A]
     with LinearSeq[A]
     with LinearSeqOps[A, LazyList, LazyList[A]]
     with IterableFactoryDefaults[A, LazyList]
     with Serializable {
   import LazyList._
 
-  @volatile private[this] var stateEvaluated: Boolean = false
-  @inline private def stateDefined: Boolean = stateEvaluated
-  private[this] var midEvaluation = false
-
-  private lazy val state: State[A] = {
-    // if it's already mid-evaluation, we're stuck in an infinite
-    // self-referential loop (also it's empty)
-    if (midEvaluation) {
-      throw new RuntimeException(
-        "LazyList evaluation depends on its own result (self-reference); see docs for more info"
-      )
+  private def this(head: A, tail: LazyList[A]) = {
+    this(LazyList.EmptyFun)
+    _head = head
+    _state = tail
+  }
+
+  // used to synchronize lazy state evaluation
+  // when `_head ne Uninitialized`
+  //   - `null` if this is an empty lazy list
+  //   - `head: A` otherwise (can be `null`, `_state == null` is used to test emptyness)
+  @volatile private[this] var _head: Any = Uninitialized
+
+  // when `_head eq Uninitialized`
+  //   - `lazySate: () => State`
+  //   - MidEvaluation while evaluating lazyState
+  // when `_head ne Uninitialized`
+  //   - `null` if this is an empty lazy list
+  //   - `tail: LazyList[A]` otherwise
+  private[this] var _state: AnyRef = lazyState
+
+  private def rawHead: Any = _head
+  private def rawState: AnyRef = _state
+
+  @inline private def stateDefined: Boolean = _head.asInstanceOf[AnyRef] ne Uninitialized
```

Comment:
is it worth also checking `MidEvaluation` here? I'm trying to figure out if this could cause a change in behaviour (and whether or not that would be incorrect)

Reply:
`MidEvaluation` is only assigned to `_state`, the `_head` remains `Uninitialized` until initialization is done.

Reply:
actually, this is definitely wrong. as it is, it permits `initState()` to return before the state is finished being initialized when `_state eq MidEvaluation`.

what's the cost of `isInstanceOf` compared to `eq`? would it make sense for `Uninitialized` and `MidEvaluation` to both extend some sealed trait that can be tested for here?